File: webapp/src/scalarField.ts
/**
 * Scalar field: f(v) = Σ c_k · φ_k(v) = coefficients @ eigenvectors
 */
export function computeScalarField(
  eigenvectors: number[][],
  coefficients: number[]
): number[] {
  const N = Math.min(eigenvectors.length, coefficients.length);
  const V = eigenvectors[0]?.length ?? 0;
  const out: number[] = new Array(V).fill(0);
  for (let v = 0; v < V; v++) {
    let s = 0;
    for (let i = 0; i < N; i++) s += coefficients[i] * (eigenvectors[i][v] ?? 0);
    out[v] = s;
  }
  return out;
}

File: analysis/scalar_field.py
"""
Scalar field synthesis: f(v) = Σ a_k · φ_k(v) = coefficients @ eigenvectors.
eigenvectors (N, V), coefficients (N,) -> scalar_values (V,).
"""
import numpy as np


def compute_scalar_field(eigenvectors, coefficients):
    """
    eigenvectors: (N, V), coefficients: (N,).
    Returns: (V,) float.
    """
    e = np.asarray(eigenvectors, dtype=np.float64)
    c = np.asarray(coefficients, dtype=np.float64).ravel()
    if e.shape[0] != len(c):
        raise ValueError("eigenvectors.shape[0] != len(coefficients)")
    return (c @ e).astype(np.float64)
